plugins/reporters/trustsource/src/main/kotlin/TrustSourceReporter.kt

```diff
@@ -0,0 +1,102 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.ossreviewtoolkit.plugins.reporters.trustsource
+
+import java.io.File
+
+import kotlinx.serialization.json.Json
+import kotlinx.serialization.json.encodeToStream
+
+import org.ossreviewtoolkit.model.DependencyNode
+import org.ossreviewtoolkit.model.licenses.LicenseView
+import org.ossreviewtoolkit.reporter.Reporter
+import org.ossreviewtoolkit.reporter.ReporterInput
+
+class TrustSourceReporter : Reporter {
+    companion object {
+        val JSON = Json { encodeDefaults = false }
+    }
+
+    override val type = "TrustSource"
+
+    private val reportFilename = "trustsource-report.json"
+
+    override fun generateReport(
+        input: ReporterInput,
+        outputDir: File,
+        options: Map<String, String>
+    ): List<File> {
+        val outputFile = outputDir.resolve(reportFilename)
+
+        val nav = input.ortResult.dependencyNavigator
+        val modules = input.ortResult.getProjects().map { project ->
+            fun traverseDeps(deps: Sequence<DependencyNode>): List<TrustSourceDependency> {
+                val tsDeps = deps.map { dep ->
+                    val pkg = input.ortResult.getPackage(dep.id)
+
+                    val resolvedLicenseInfo = input.licenseInfoResolver.resolveLicenseInfo(dep.id).filterExcluded()
+                    val effectiveLicense = resolvedLicenseInfo.effectiveLicense(
+                        LicenseView.CONCLUDED_OR_DECLARED_AND_DETECTED,
+                        input.ortResult.getPackageLicenseChoices(dep.id),
+                        input.ortResult.getRepositoryLicenseChoices()
+                    )
+                    val licenses = effectiveLicense?.decompose()?.map {
+                        val name = it.toString()
+                        val url = it.getLicenseUrl() ?: ""
+
+                        TrustSourceLicense(name, url)
+                    }
+
+                    TrustSourceDependency(
+                        key = "${dep.id.type}:${dep.id.name}",
+                        name = dep.id.name,
+                        repoUrl = pkg?.metadata?.sourceArtifact?.url ?: "",
+                        homepageUrl = pkg?.metadata?.homepageUrl ?: "",
+                        description = pkg?.metadata?.description ?: "",
+                        checksum = "",
+                        private = false,
+
+                        versions = listOf(dep.id.version),
+
+                        dependencies = dep.visitDependencies { traverseDeps(it) },
+                        licenses = licenses ?: emptyList(),
+                        meta = TrustSourceMeta()
+                    )
+                }
+
+                return tsDeps.toList()
+            }
```

Comment:
This is a bit large for a local function Better extract it to a private top-level function outside of the class, if possible.